src/USER-MISC/pair_agni.cpp

```diff
@@ -347,132 +358,123 @@ double PairAGNI::init_one(int i, int j)
+      values.next_string(); // ignore
+      nparams = values.next_int();
 
-    // words = ptrs to all words in line
-
-    nwords = 0;
-    words[nwords++] = strtok(line," \t\n\r\f");
-    while ((words[nwords++] = strtok(nullptr," \t\n\r\f"))) continue;
-    --nwords;
-
-    if ((nwords == 2) && (strcmp(words[0],"generation") == 0)) {
-      int ver = atoi(words[1]);
-      if (ver != AGNI_VERSION)
-        error->all(FLERR,"Incompatible AGNI potential file version");
-      if ((ver == 1) && (nelements != 1))
-        error->all(FLERR,"Cannot handle multi-element systems with this potential");
-
-    } else if ((nwords == 2) && (strcmp(words[0],"n_elements") == 0)) {
-      nparams = atoi(words[1]);
       if ((nparams < 1) || params) // sanity check
         error->all(FLERR,"Invalid AGNI potential file");
       params = memory->create(params,nparams,"pair:params");
       memset(params,0,nparams*sizeof(Param));
+      
       curparam = -1;
-
-    } else if (params && (nwords == nparams+1) && (strcmp(words[0],"element") == 0)) {
       wantdata = -1;
+
+      values = reader.next_values(2);
+      values.next_string(); // ignore
       for (i = 0; i < nparams; ++i) {
+        std::string element = values.next_string();
         for (j = 0; j < nelements; ++j)
-          if (strcmp(words[i+1],elements[j]) == 0) break;
+          if (strcmp(element.c_str(),elements[j]) == 0) break;
         if (j == nelements)
           error->all(FLERR,"No suitable parameters for requested element found");
         else params[i].ielement = j;
       }
-    } else if (params && (nwords == 2) && (strcmp(words[0],"interaction") == 0)) {
-      for (i = 0; i < nparams; ++i)
-        if (strcmp(words[1],elements[params[i].ielement]) == 0) curparam = i;
-    } else if ((curparam >=0) && (nwords == 1) && (strcmp(words[0],"endVar") == 0)) {
-      int numtrain = params[curparam].numtrain;
-      int numeta = params[curparam].numeta;
-      params[curparam].alpha = new double[numtrain];
-      params[curparam].yU = new double[numtrain];
-      params[curparam].xU = new double*[numeta];
-      for (i = 0; i < numeta; ++i)
-        params[curparam].xU[i] = new double[numtrain];
 
+      values = reader.next_values(2);
+      values.next_string(); // ignore
+      for (i = 0; i < nparams; ++i){
+        std::string element = values.next_string();
+        if (strcmp(element.c_str(),elements[params[i].ielement]) == 0) curparam = i;
+      }
+
+      values = reader.next_values(2);
+      values.next_string(); // ignore
+      atomic_feature_version = values.next_int();
+      if (atomic_feature_version != AGNI_VERSION_1 && atomic_feature_version != AGNI_VERSION_2)
+        error->all(FLERR,"Incompatible AGNI potential file version");
+
+      values = reader.next_values(2);
+      values.next_string(); // ignore
+      params[curparam].numeta = values.next_int();
+      params[curparam].eta = new double[params[curparam].numeta];
```

Comment:
Here is a parallel programming problem. This part of the code only runs on MPI rank 0, so the memory is only allocated on that rank. And even if it was allocated on all MPI ranks, you cannot broadcast the pointer to all MPI ranks since "new" or malloc() will return different addresses on different MPI ranks. Thus this array and the `xU` one in the next line need to be handled special. You can keep the code as is, but after the broadcast of the struct you need to repeat the allocations with "new" on all MPI ranks != 0 and then broadcast the corresponding arrays individually.